From 43c17e62dbffdd64eaf1a69307b26c00fb19cdff Mon Sep 17 00:00:00 2001
From: yvanliang <yvan6liang@outlook.com>
Date: Fri, 9 Oct 2026 15:43:17 +0800
Subject: [PATCH 4/9] GSI Android 17 compatibility:
 0004-Don-t-abort-when-failing-to-get-real-time-priority.patch

Change-Id: I69cc2998b5b40a869d92083a0fe6a1a0713d3931
---
 system/btif/src/btif_a2dp_sink.cc   |  2 +-
 system/btif/src/btif_a2dp_source.cc |  2 +-
 system/osi/src/alarm.cc             | 24 +++++++++++++++++-------
 system/stack/btu/main_thread.cc     |  4 ----
 4 files changed, 19 insertions(+), 13 deletions(-)

diff --git a/system/btif/src/btif_a2dp_sink.cc b/system/btif/src/btif_a2dp_sink.cc
index e329f6c..c12ea75 100644
--- a/system/btif/src/btif_a2dp_sink.cc
+++ b/system/btif/src/btif_a2dp_sink.cc
@@ -170,7 +170,7 @@ bool btif_a2dp_sink_init() {
   /* Schedule the rest of the operations */
   if (!btif_a2dp_sink_cb.worker_thread.EnableRealTimeScheduling()) {
 #if defined(__ANDROID__)
-    log::fatal("Failed to increase A2DP decoder thread priority");
+    log::error("Failed to increase A2DP decoder thread priority");
 #endif
   }
   btif_a2dp_sink_cb.worker_thread.DoInThread(base::BindOnce(btif_a2dp_sink_init_delayed));
diff --git a/system/btif/src/btif_a2dp_source.cc b/system/btif/src/btif_a2dp_source.cc
index d9a1402..b903069 100644
--- a/system/btif/src/btif_a2dp_source.cc
+++ b/system/btif/src/btif_a2dp_source.cc
@@ -325,7 +325,7 @@ bool btif_a2dp_source_init(void) {
 
   if (!btif_a2dp_source_thread.EnableRealTimeScheduling()) {
 #if defined(__ANDROID__)
-    log::fatal("unable to enable real time scheduling");
+    log::error("unable to enable real time scheduling");
 #endif
   }
 
diff --git a/system/osi/src/alarm.cc b/system/osi/src/alarm.cc
index f3468af..6ace2bc 100644
--- a/system/osi/src/alarm.cc
+++ b/system/osi/src/alarm.cc
@@ -685,17 +685,27 @@ static bool timer_create_internal(const clockid_t clock_id, timer_t* timer) {
   sigevent.sigev_notify = SIGEV_THREAD;
   sigevent.sigev_notify_function = timer_callback;
   sigevent.sigev_notify_attributes = &thread_attr;
-  if (timer_create(clock_id, &sigevent, timer) == -1) {
+
+  int ret = timer_create(clock_id, &sigevent, timer);
+  if (ret == -1) {
+    log::error("{} failed to create timer with RT err {}... Try again without RT", __func__, strerror(errno));
+    // Recreate timer without RT priority
+    memset(&sigevent, 0, sizeof(sigevent));
+    sigevent.sigev_notify = SIGEV_THREAD;
+    sigevent.sigev_notify_function = (void (*)(union sigval))timer_callback;
+    ret = timer_create(clock_id, &sigevent, timer);
+  }
+  if (ret == -1) {
     log::error("unable to create timer with clock {}: {}", clock_id, strerror(errno));
     if (clock_id == CLOCK_BOOTTIME_ALARM) {
       log::error(
-              "The kernel might not have support for "
-              "timer_create(CLOCK_BOOTTIME_ALARM): "
-              "https://lwn.net/Articles/429925/");
+        "The kernel might not have support for "
+        "timer_create(CLOCK_BOOTTIME_ALARM): "
+        "https://lwn.net/Articles/429925/");
       log::error(
-              "See following patches: "
-              "https://git.kernel.org/cgit/linux/kernel/git/torvalds/linux.git/log/"
-              "?qt=grep&q=CLOCK_BOOTTIME_ALARM");
+        "See following patches: "
+        "https://git.kernel.org/cgit/linux/kernel/git/torvalds/linux.git/log/"
+        "?qt=grep&q=CLOCK_BOOTTIME_ALARM");
     }
     return false;
   }
diff --git a/system/stack/btu/main_thread.cc b/system/stack/btu/main_thread.cc
index 9bd84ae..2d48693 100644
--- a/system/stack/btu/main_thread.cc
+++ b/system/stack/btu/main_thread.cc
@@ -70,11 +70,7 @@ void main_thread_start_up() {
     log::fatal("unable to start btu message loop thread.");
   }
   if (!main_thread.EnableRealTimeScheduling()) {
-#if defined(__ANDROID__)
-    log::fatal("unable to enable real time scheduling");
-#else
     log::error("unable to enable real time scheduling");
-#endif
   }
 }
 
